File: rust/2021/src/bin/day_06.rs
use std::fs;
use aoc::counter::Counter;

fn simulate(nums: &[u8], days: usize) -> usize {
    let mut fish_counter: Counter<_> = nums.iter().cloned().collect();
    for _ in 0..days {
        let previous = fish_counter;
        fish_counter = Counter::new();
        for (timer, count) in previous {
            if timer > 0 {
                fish_counter[&(timer - 1)] += count;
            } else {
                fish_counter[&6] += count;
                fish_counter[&8] = count;
            }
        }
    }
    fish_counter.values().sum()
}

fn main() {
    let input = fs::read_to_string("../input/2021/input_06.txt").unwrap();
    let nums: Vec<_> = input.trim().split(',').map(|s| s.parse().unwrap()).collect();
    println!("Part 1: {}", simulate(&nums, 80));
    println!("Part 2: {}", simulate(&nums, 256));
}

File: haskell/2021/day_06.hs
import qualified Data.List.Split as Split
import qualified Data.Map as Map

step :: Map.Map Int Int -> Map.Map Int Int
step = Map.fromListWith (+) . concatMap (\(timer, count) ->
        if timer > 0 then [(timer - 1, count)] else [(6, count), (8, count)]
    ) . Map.assocs

main :: IO ()
main = do
    text <- readFile "../../input/2021/input_06.txt"
    let timers = read <$> Split.splitOn "," (head $ words text)
    let fish = Map.fromListWith (+) $ zip timers (repeat 1)
    putStrLn $ "Part 1: " ++ show (sum $ iterate step fish !! 80)
    putStrLn $ "Part 2: " ++ show (sum $ iterate step fish !! 256)
